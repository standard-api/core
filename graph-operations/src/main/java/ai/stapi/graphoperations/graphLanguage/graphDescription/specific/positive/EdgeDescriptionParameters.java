package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class EdgeDescriptionParameters implements GraphDescriptionParameters {

  private String edgeType;

  private EdgeDescriptionParameters() {
  }

  public EdgeDescriptionParameters(String edgeType) {
    this.edgeType = edgeType;
  }

  public String getEdgeType() {
    return edgeType;
  }
}
