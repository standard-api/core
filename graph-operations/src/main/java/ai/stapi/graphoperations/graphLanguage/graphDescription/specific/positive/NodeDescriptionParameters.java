package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;

public class NodeDescriptionParameters implements GraphDescriptionParameters {

  private String nodeType;

  private NodeDescriptionParameters() {

  }

  public NodeDescriptionParameters(String nodeType) {
    this.nodeType = nodeType;
  }

  public String getNodeType() {
    return nodeType;
  }
}