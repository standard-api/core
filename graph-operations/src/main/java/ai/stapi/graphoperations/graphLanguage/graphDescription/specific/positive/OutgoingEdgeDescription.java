package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.List;

public class OutgoingEdgeDescription extends AbstractEdgeDescription {

  public static final String SERIALIZATION_TYPE = "OutgoingEdgeDescription";

  protected OutgoingEdgeDescription() {

  }

  public OutgoingEdgeDescription(
      EdgeDescriptionParameters parameters,
      List<GraphDescription> childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
  }

  public OutgoingEdgeDescription(
      EdgeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
  }

  public OutgoingEdgeDescription(EdgeDescriptionParameters parameters) {
    super(parameters, SERIALIZATION_TYPE);
  }

  @Override
  public boolean isOutgoing() {
    return true;
  }
}
