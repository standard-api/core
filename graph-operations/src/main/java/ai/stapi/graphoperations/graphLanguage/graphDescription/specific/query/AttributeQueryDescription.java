package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractAttributeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AttributeDescriptionParameters;
import java.util.ArrayList;
import java.util.List;

public class AttributeQueryDescription extends AbstractAttributeDescription implements QueryGraphDescription {

  public static final String SERIALIZATION_TYPE = "AttributeQueryDescription";
  public static final String DESCRIBED_STRUCTURE_TYPE = "Irrelevant";

  private AttributeQueryDescription() {
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      GraphDescription... childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      AttributeDescriptionParameters parameters,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, parameters, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName,
      List<GraphDescription> childDescriptions
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, childDescriptions);
  }

  public AttributeQueryDescription(
      String attributeName
  ) {
    super(SERIALIZATION_TYPE, DESCRIBED_STRUCTURE_TYPE, attributeName, new ArrayList<>());
  }
}
