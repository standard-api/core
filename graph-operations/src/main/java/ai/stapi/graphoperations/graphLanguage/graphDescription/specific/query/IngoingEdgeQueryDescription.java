package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import java.util.List;

public class IngoingEdgeQueryDescription extends AbstractEdgeDescription
    implements EdgeQueryDescription {

  public static final String SERIALIZATION_TYPE = "16843ef0-b5a1-4d45-b59c-962294de5e6e";

  private SearchQueryParameters searchQueryParameters;

  private boolean isCompact;
  private CollectionComparisonOperator collectionComparisonOperator;

  protected IngoingEdgeQueryDescription() {
    super();
  }

  public IngoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      CollectionComparisonOperator collectionComparisonOperator,
      List<GraphDescription> childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
    this.collectionComparisonOperator = collectionComparisonOperator;
    this.isCompact = true;
  }

  public IngoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      List<GraphDescription> childDeclarations
  ) {
    this(parameters, searchQueryParameters, CollectionComparisonOperator.ANY, childDeclarations);
  }

  public IngoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      CollectionComparisonOperator collectionComparisonOperator,
      GraphDescription... childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
    this.collectionComparisonOperator = collectionComparisonOperator;
    this.isCompact = true;
  }

  public IngoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    this(parameters, searchQueryParameters, CollectionComparisonOperator.ANY, childDeclarations);
  }

  public IngoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters
  ) {
    super(parameters, SERIALIZATION_TYPE);
    this.searchQueryParameters = new SearchQueryParameters();
    this.isCompact = true;
  }

  public static IngoingEdgeQueryDescription asConnections(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    var ingoingEdgeQueryDescription = new IngoingEdgeQueryDescription(
        parameters,
        searchQueryParameters,
        childDeclarations
    );
    ingoingEdgeQueryDescription.isCompact = false;
    return ingoingEdgeQueryDescription;
  }

  public static IngoingEdgeQueryDescription asConnections(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      CollectionComparisonOperator collectionComparisonOperator,
      GraphDescription... childDeclarations
  ) {
    var ingoingEdgeQueryDescription = new IngoingEdgeQueryDescription(
        parameters,
        searchQueryParameters,
        collectionComparisonOperator,
        childDeclarations
    );
    ingoingEdgeQueryDescription.isCompact = false;
    return ingoingEdgeQueryDescription;
  }

  @Override
  public boolean isOutgoing() {
    return false;
  }

  @Override
  public SearchQueryParameters getSearchQueryParameters() {
    return this.searchQueryParameters;
  }

  @Override
  public boolean isCompact() {
    return this.isCompact;
  }

  public void setIsCompact(boolean isCompact) {
    this.isCompact = isCompact;
  }

  public CollectionComparisonOperator getCollectionComparisonOperator() {
    return collectionComparisonOperator;
  }
}
