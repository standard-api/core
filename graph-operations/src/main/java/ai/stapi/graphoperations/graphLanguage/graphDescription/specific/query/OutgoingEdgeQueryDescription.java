package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import java.util.List;

public class OutgoingEdgeQueryDescription extends AbstractEdgeDescription
    implements EdgeQueryDescription {

  public static final String SERIALIZATION_TYPE = "399c927c-fa45-435d-bc83-66918b18482c";

  private SearchQueryParameters searchQueryParameters;
  private CollectionComparisonOperator collectionComparisonOperator;
  private boolean isCompact;

  protected OutgoingEdgeQueryDescription() {
    super();
  }

  public OutgoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      CollectionComparisonOperator collectionComparisonOperator,
      List<GraphDescription> childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
    this.collectionComparisonOperator = collectionComparisonOperator;
    this.isCompact = true;
  }

  public OutgoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      List<GraphDescription> childDeclarations
  ) {
    this(parameters, searchQueryParameters, CollectionComparisonOperator.ANY, childDeclarations);
  }

  public OutgoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      CollectionComparisonOperator collectionComparisonOperator,
      GraphDescription... childDeclarations
  ) {
    super(parameters, SERIALIZATION_TYPE, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
    this.collectionComparisonOperator = collectionComparisonOperator;
    this.isCompact = true;
  }

  public OutgoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    this(parameters, searchQueryParameters, CollectionComparisonOperator.ANY, childDeclarations);
  }

  public OutgoingEdgeQueryDescription(
      EdgeDescriptionParameters parameters
  ) {
    super(parameters, SERIALIZATION_TYPE);
    this.searchQueryParameters = new SearchQueryParameters();
    this.isCompact = true;
  }

  public static OutgoingEdgeQueryDescription asConnections(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    var outgoingEdgeQueryDescription = new OutgoingEdgeQueryDescription(
        parameters,
        searchQueryParameters,
        childDeclarations
    );
    outgoingEdgeQueryDescription.isCompact = false;
    return outgoingEdgeQueryDescription;
  }

  public static OutgoingEdgeQueryDescription asConnections(
      EdgeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      CollectionComparisonOperator collectionComparisonOperator,
      GraphDescription... childDeclarations
  ) {
    var outgoingEdgeQueryDescription = new OutgoingEdgeQueryDescription(
        parameters,
        searchQueryParameters,
        collectionComparisonOperator,
        childDeclarations
    );
    outgoingEdgeQueryDescription.isCompact = false;
    return outgoingEdgeQueryDescription;
  }


  @Override
  public boolean isOutgoing() {
    return true;
  }

  @Override
  public SearchQueryParameters getSearchQueryParameters() {
    return this.searchQueryParameters;
  }

  @Override
  public boolean isCompact() {
    return isCompact;
  }

  public void setIsCompact(boolean isCompact) {
    this.isCompact = isCompact;
  }

  public CollectionComparisonOperator getCollectionComparisonOperator() {
    return collectionComparisonOperator;
  }
}
