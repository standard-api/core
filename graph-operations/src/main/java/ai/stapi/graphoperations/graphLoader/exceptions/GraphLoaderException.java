package ai.stapi.graphoperations.graphLoader.exceptions;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.utils.Stringifier;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import ai.stapi.identity.UniqueIdentifier;
import java.util.List;

public class GraphLoaderException extends RuntimeException {

  public GraphLoaderException(String becauseMessage) {
    super("Cannot load graph, " + becauseMessage);
  }

  private GraphLoaderException(String becauseMessage, Throwable cause) {
    super("Cannot load graph, " + becauseMessage, cause);
  }

  public static GraphLoaderException becauseThereIsNoGraphElementWithProvidedUuid(
      UniqueIdentifier uuid,
      GraphDescription graphDescription
  ) {
    return new GraphLoaderException(
        "because specified graph element does not exist in graph." +
            "\nUUID: " + uuid +
            "\nGraphDescription: " + graphDescription.getClass().getSimpleName() +
            "\n" + Stringifier.convertToString(graphDescription)
    );
  }

  public static GraphLoaderException becauseThereIsMoreThanOneNodeDescriptionsOnEdgeDescription(
      UniqueIdentifier id, String type) {
    return new GraphLoaderException(
        "because there is more than one node descriptions on edge description." +
            "\nUnique Identifier: " + id +
            "\nEdge type: " + type
    );
  }

  public static GraphLoaderException becauseEdgeDescriptionHadMoreChildNodeDescriptions(
      AbstractEdgeDescription edgeDescription,
      List<String> nodeTypes
  ) {
    var param = (EdgeDescriptionParameters) edgeDescription.getParameters();
    return new GraphLoaderException(
        "because edge description had more child node descriptions." +
            "\nEdge Description ClassName: " + edgeDescription.getClass().getSimpleName() +
            "\nEdge type: " + param.getEdgeType() +
            "\nNode Types: " + String.join(", ", nodeTypes)
    );
  }

  public static GraphLoaderException becauseProvidedOGMDidNotStartWithObjectOGM(
      ObjectGraphMapping objectGraphMapping) {
    return new GraphLoaderException(
        "because provided OGM did not start with object OGM." +
            "\nOGM Class: " + objectGraphMapping.getClass().getSimpleName()
    );
  }
}
