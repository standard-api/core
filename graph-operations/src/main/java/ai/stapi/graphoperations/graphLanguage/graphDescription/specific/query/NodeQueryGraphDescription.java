package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.AbstractNodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import java.util.List;

public class NodeQueryGraphDescription extends AbstractNodeDescription
    implements GraphElementQueryDescription {

  public static final String SERIALIZATION_TYPE = "72ba0bef-5f38-489b-bbbf-8212191f1f92";

  private SearchQueryParameters searchQueryParameters;

  protected NodeQueryGraphDescription() {
    super();
  }

  public NodeQueryGraphDescription(NodeDescription nodeDescription) {
    super(
        SERIALIZATION_TYPE,
        (NodeDescriptionParameters) nodeDescription.getParameters(),
        nodeDescription.getChildGraphDescriptions()
    );
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      List<GraphDescription> childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.parameters = parameters;
    this.searchQueryParameters = searchQueryParameters;
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.parameters = parameters;
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters
  ) {
    super(SERIALIZATION_TYPE, parameters);
    this.searchQueryParameters = new SearchQueryParameters();
  }

  public NodeQueryGraphDescription(
      NodeDescriptionParameters parameters,
      SearchQueryParameters searchQueryParameters,
      GraphDescription... childDeclarations
  ) {
    super(SERIALIZATION_TYPE, parameters, childDeclarations);
    this.searchQueryParameters = searchQueryParameters;
  }

  @Override
  public SearchQueryParameters getSearchQueryParameters() {
    return this.searchQueryParameters;
  }
}
