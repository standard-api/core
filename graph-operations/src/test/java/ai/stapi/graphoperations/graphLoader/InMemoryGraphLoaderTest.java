package ai.stapi.graphoperations.graphLoader;

import ai.stapi.graph.Graph;
import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.EdgeDescriptionParameters;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NodeDescriptionParameters;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.OutgoingEdgeDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.AttributeQueryDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.IngoingEdgeQueryDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.NodeQueryGraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.query.OutgoingEdgeQueryDescription;
import ai.stapi.graphoperations.graphLoader.exceptions.GraphLoaderException;
import ai.stapi.graphoperations.graphLoader.inmemory.InMemoryGenericSearchOptionResolver;
import ai.stapi.graphoperations.graphLoader.inmemory.InMemoryGraphLoader;
import ai.stapi.graphoperations.graphLoader.search.SearchQueryParameters;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import ai.stapi.test.schemaintegration.SchemaIntegrationTestCase;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

class InMemoryGraphLoaderTest extends SchemaIntegrationTestCase {
  
  @Autowired
  private InMemoryGenericSearchOptionResolver searchOptionResolver;
  
  @Autowired
  private ObjectMapper objectMapper;
  
  @Autowired
  private StructureSchemaFinder structureSchemaFinder;

  @Test
  void FindMethod_ShouldReturnEmptyListWhenGraphIsEmpty_WhenSearchingNode() {
    var graphLoader = this.initializeGraphLoader(this.getEmptyGraph());
    var nodeDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("any_node")
    );
    var result = graphLoader.findAsTraversable(nodeDescription);
    Assertions.assertEquals(result.size(), 0);
  }

  @Test
  void FindMethod_ShouldReturnEmptyListWhenGraphIsEmpty_WhenSearchingOutgoingEdge() {
    var graphLoader = this.initializeGraphLoader(this.getEmptyGraph());
    var edgeDescription = new OutgoingEdgeQueryDescription(
        new EdgeDescriptionParameters("any_edge")
    );
    var result = graphLoader.findAsTraversable(edgeDescription);
    Assertions.assertEquals(result.size(), 0);
  }

  @Test
  void FindMethod_ShouldReturnEmptyListWhenGraphIsEmpty_WhenSearchingIngoingEdge() {
    var graphLoader = this.initializeGraphLoader(this.getEmptyGraph());
    var edgeDescription = new IngoingEdgeQueryDescription(
        new EdgeDescriptionParameters("any_edge")
    );
    var result = graphLoader.findAsTraversable(edgeDescription);
    Assertions.assertEquals(result.size(), 0);
  }

  @Test
  void FindMethod_shouldReturnNodesSpecifiedByFirstNodeQueryDescription() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node1")
    );
    var result = graphLoader.findAsTraversable(nodeQueryDescription);
    this.thenGraphElementsApproved(result);
  }

  @Test
  void FindMethod_shouldReturnNodesWithAttributesAsSpecified() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node1"),
        new SearchQueryParameters(),
        new AttributeQueryDescription("example_attribute")
    );
    var result = graphLoader.findAsTraversable(nodeQueryDescription);
    this.thenGraphElementsApproved(result);
  }

  @Test
  void FindMethod_shouldReturnNodesEvenThoughSpecifiedAttributeNotPresent() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node1"),
        new SearchQueryParameters(),
        new AttributeQueryDescription("not_existing_attribute")
    );
    var result = graphLoader.findAsTraversable(nodeQueryDescription);
    this.thenGraphElementsApproved(result);
  }

  @Test
  void FindMethod_shouldReturnNodesWithSpecifiedAttributesEvenThoughSomeAreMissing() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node1"),
        new SearchQueryParameters(),
        new AttributeQueryDescription("example_attribute"),
        new AttributeQueryDescription("other_attribute")
    );
    var result = graphLoader.findAsTraversable(nodeQueryDescription);
    this.thenGraphElementsApproved(result);
  }

  @Test
  void FindMethod_shouldReturnNodesWithSubgraphAsSpecified() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node2"),
        new OutgoingEdgeDescription(
            new EdgeDescriptionParameters("edge2"),
            new NodeQueryGraphDescription(
                new NodeDescriptionParameters("node3"),
                new AttributeQueryDescription("example_attribute")
            )
        )
    );
    var result = graphLoader.find(
        nodeQueryDescription,
        Object.class,
        GraphLoaderReturnType.GRAPH
    ).getGraphLoaderFindAsGraphOutput();

    this.thenGraphApproved(result.getGraph());

    Assertions.assertEquals(result.getFoundGraphElementIds().size(), 1);
    Assertions.assertEquals(
        result.getFoundGraphElementIds().get(0),
        UniversallyUniqueIdentifier.fromString("2378d6d6-4424-45f7-8f2e-b21bf5e8de66")
    );
  }

  @Test
  void GetMethod_shouldThrowErrorWhenNodeDoesNotExistWithSpecifiedUUID() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node2")
    );

    Executable executable = () -> graphLoader.getAsTraversable(
        UniversallyUniqueIdentifier.fromString("0762fc65-7c9f-4578-944a-8d89f003d03d"),
        nodeQueryDescription
    );
    this.thenExceptionMessageApproved(GraphLoaderException.class, executable);
  }

  @Test
  void GetMethod_shouldThrowErrorWhenEdgeDoesNotExistWithSpecifiedUUID() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new OutgoingEdgeQueryDescription(
        new EdgeDescriptionParameters("nonexisitng_edge")
    );

    Executable executable = () -> graphLoader.getAsTraversable(
        UniversallyUniqueIdentifier.fromString("0762fc65-7c9f-4578-944a-8d89f003d03d"),
        nodeQueryDescription
    );
    Assertions.assertThrows(GraphLoaderException.class, executable);
  }

  @Test
  void GetMethod_shouldReturnSubgraphAsSpecified() {
    var graphLoader = this.initializeGraphLoader(this.getSimpleGraph());
    var nodeQueryDescription = new NodeQueryGraphDescription(
        new NodeDescriptionParameters("node1"),
        new AttributeQueryDescription("example_attribute"),
        new OutgoingEdgeDescription(
            new EdgeDescriptionParameters("edge1"),
            new NodeQueryGraphDescription(
                new NodeDescriptionParameters("node2"),
                new AttributeQueryDescription("example_attribute"),
                new OutgoingEdgeDescription(
                    new EdgeDescriptionParameters("edge2"),
                    new NodeQueryGraphDescription(
                        new NodeDescriptionParameters("node3"),
                        new AttributeQueryDescription("example_attribute")
                    )
                )
            )
        )
    );
    var result = graphLoader.get(
        UniversallyUniqueIdentifier.fromString("fca128a0-637a-4d6f-bd86-d2e42fe9e38d"),
        nodeQueryDescription,
        Object.class,
        GraphLoaderReturnType.GRAPH
    ).getGraph();
    this.thenGraphApproved(result);
  }

  private Graph getSimpleGraph() {
    var graph = new Graph();
    var node1 = new Node(
        UniversallyUniqueIdentifier.fromString("fca128a0-637a-4d6f-bd86-d2e42fe9e38d"), "node1");
    node1 = node1.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("node1_value")));
    node1 = node1.add(
        new LeafAttribute<>("other_attribute", new StringAttributeValue("node1_other_value")));
    var node11 = new Node(
        UniversallyUniqueIdentifier.fromString("376a6385-60bb-4cbb-97f2-6b8f5349fb8e"), "node1");
    node11 = node11.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("node11_value")));
    var node12 = new Node(
        UniversallyUniqueIdentifier.fromString("b0c2711f-aa50-4227-8c83-353749c872de"), "node1");
    node12 = node12.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("node12_value")));
    var node2 = new Node(
        UniversallyUniqueIdentifier.fromString("2378d6d6-4424-45f7-8f2e-b21bf5e8de66"), "node2");
    node2 = node2.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("node2_value")));
    var node3 = new Node(
        UniversallyUniqueIdentifier.fromString("25ed70b7-c53f-4d5a-b279-2595ad2c1f4c"), "node3");
    node3 = node3.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("node3_value")));
    var edge1 = new Edge(
        UniversallyUniqueIdentifier.fromString("b18de82b-7c69-46e8-b18b-f45622df3459"), node1,
        "edge1", node2);
    edge1 = edge1.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("edge1_value")));
    var edge2 = new Edge(
        UniversallyUniqueIdentifier.fromString("d3cba05a-7918-4ebc-8bd1-137dee05a0c5"), node2,
        "edge2", node3);
    edge2 = edge2.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("edge2_value")));
    var edge3 = new Edge(
        UniversallyUniqueIdentifier.fromString("42e380ce-9794-49f0-9380-c2ff275b9bfa"), node3,
        "edge3", node1);
    edge3 = edge3.add(
        new LeafAttribute<>("example_attribute", new StringAttributeValue("edge3_value")));
    return graph.withAll(node1, node11, node12, node2, node3, edge1, edge2, edge3);
  }

  private Graph getEmptyGraph() {
    return new Graph();
  }

  private InMemoryGraphLoader initializeGraphLoader(Graph graph) {
    return new InMemoryGraphLoader(
        graph, 
        this.searchOptionResolver, 
        this.structureSchemaFinder, 
        this.objectMapper
    );
  }
}
