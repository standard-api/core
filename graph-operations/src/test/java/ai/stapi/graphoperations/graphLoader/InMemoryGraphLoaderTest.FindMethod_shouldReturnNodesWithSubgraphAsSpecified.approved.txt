nodes:
	node_type: node2
	node_hash: A6D3E
	node_edges:
		node2 -> edge2 -> node3 (A6D3E -> E50BE -> 586A7)
	attributes:

	node_type: node3
	node_hash: 586A7
	node_edges:
		node2 -> edge2 -> node3 (A6D3E -> E50BE -> 586A7)
	attributes:
		example_attribute -> node3_value

edges:
	edge_type: edge2
	edge_hash: E50BE
	node2 -> edge2 -> node3 (A6D3E -> E50BE -> 586A7)
	attributes:
